sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/OperatorStateMetadata.scala

```diff
@@ -358,3 +358,119 @@ class OperatorStateMetadataV2Reader(
+  private val stateSchemaPath = stateStoreWriter.stateSchemaDirPath()
+  private val metadataDirPath = OperatorStateMetadataV2.metadataDirPath(stateOpIdPath)
+  private lazy val fm = CheckpointFileManager.create(metadataDirPath, hadoopConf)
+
+  protected def isBatchFile(path: Path): Boolean = {
+    try {
+      path.getName.toLong
+      true
+    } catch {
+      case _: NumberFormatException => false
+    }
+  }
+
+  /**
+   * A `PathFilter` to filter only batch files
+   */
+  protected val batchFilesFilter: PathFilter = (path: Path) => isBatchFile(path)
+
+  private def pathToBatchId(path: Path): Long = {
+    path.getName.toLong
+  }
+
+  def purgeMetadataFiles(): Unit = {
+    val thresholdBatchId = findThresholdBatchId()
+    if (thresholdBatchId != -1) {
+      val earliestBatchIdKept = deleteMetadataFiles(thresholdBatchId)
+      // we need to delete everything from 0 to (earliestBatchIdKept - 1), inclusive
+      deleteSchemaFiles(earliestBatchIdKept - 1)
+    }
+  }
+
+  // We only want to keep the metadata and schema files for which the commit
+  // log is present, so we will delete any file that precedes the batch for the oldest
+  // commit log
+  private def findThresholdBatchId(): Long = {
+    commitLog.listBatchesOnDisk.headOption.getOrElse(0L) - 1L
+  }
+
+  private def deleteSchemaFiles(thresholdBatchId: Long): Unit = {
+    val schemaFiles = fm.list(stateSchemaPath).sorted.map(_.getPath)
+
+    if (schemaFiles.length > 1) {
+      val filesBeforeThreshold = schemaFiles.filter { path =>
+        val batchIdInPath = path.getName.split("_").head.toLong
+        batchIdInPath <= thresholdBatchId
+      }
+      filesBeforeThreshold.foreach { path =>
+        fm.delete(path)
+      }
+    }
+  }
+
+  // Deletes all metadata files that are below a thresholdBatchId, except
+  // for the latest metadata file so that we have at least 1 metadata and schema
+  // file at all times per each stateful query
+  // Returns the batchId of the earliest schema file we want to keep
+  private def deleteMetadataFiles(thresholdBatchId: Long): Long = {
+    val metadataFiles = fm.list(metadataDirPath, batchFilesFilter)
+
+    if (metadataFiles.isEmpty) {
+      return -1L // No files to delete
+    }
+
+    val sortedBatchIds = metadataFiles.map(file => pathToBatchId(file.getPath)).sorted
+    val latestBatchId = sortedBatchIds.last
+    var highestDeletedBatchId = -1L
+
+    metadataFiles.foreach { batchFile =>
+      val batchId = pathToBatchId(batchFile.getPath)
+      if (batchId <= thresholdBatchId && batchId < latestBatchId) {
+        fm.delete(batchFile.getPath)
+        highestDeletedBatchId = math.max(highestDeletedBatchId, batchId)
+      }
+    }
+    // Find the next batch id immediately greater than threshold batchId.
+    // We use this to find the metadata and schema files we want to keep
+    val nextBatchId = sortedBatchIds.find(_ > thresholdBatchId).getOrElse(latestBatchId)
```

Comment:
We should just use `latestBatchId` here right ?

Reply:
latestBatchId is the very last batchId of this list